Infer panel count in App.js from number of default threshold values

joshhead opened the pull request:
The high level goal here is to allow the React app to accommodate an unknown number of panels without code changes or rebuilding.

I tried to focus on that in this branch and not make too many unrelated changes. However, I did still make some opinionated changes.

I moved some module-level state such as kCurValues, kCurThresholds, and ws into hooks and refs. This has two benefits IMO.

First, the server can shut down and come back with a different number of sensors configured, and the app will handle this gracefully without a page reload. This is not a super important feature, but it demonstrates the second benefit, which is that more resources are managed by React and so their lifetime is well defined including how to clean them up without reloading the page.

I like this style better, but it is more complicated in some ways. There is definitely a simplicity to cleaning up state by reloading the whole page.

I tried to adopt a similar change without refactoring `connect()`. I think it works okay but I wasn't as happy with it. https://github.com/teejusb/fsr/compare/b89e5228b...joshhead:js-detect-panel-count-minimal?expand=1

joshhead commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
+function useWsConnection({ defaults, onCloseWs }) {
+  const [isWsReady, setIsWsReady] = useState(false);
+
+  const curValuesRef = useRef({
```
"curValuesRef" includes kCurValues and kCurThresholds, but it would be relatively straightforward to put them in different refs if desired. If they stay here, I'm not sure about the name.

Incidentally, what is the `k` prefix for?

teejusb replied:
The disambiguation I've been trying to use is that `Values` refers to the sensor readings and `Thresholds` are the trigger levels.

Combining is fine but I would want to rename it. Since it encapsulates all the data needed for the WebUI, maybe something like `webUIDataRef`

joshhead replied:
That naming sounds good. I like the `Ref` suffix because if you want to give a name to the current value it can easily be `const webUIData = webUIDataRef.current`.

teejusb replied:
SGTM!

joshhead commented on webui/src/App.js:
```diff
@@ -479,22 +569,13 @@ function Plot() {
-  const [profiles, setProfiles] = useState([]);
-  const [activeProfile, setActiveProfile] = useState('');
-
+function FSRWebUI(props) {
```
I renamed some components because I added a new top-level component which is now App, but I'm happy to call them something else.

teejusb replied:
This SGTM!

joshhead commented on webui/src/App.js:
```diff
@@ -575,17 +655,58 @@ function App() {
+      />
+    );
+  } else {
+    return <LoadingScreen />
```
I want to note that if the websocket closes, most of the UI is unmounted and historical sensor values are cleared, which is a change to existing behavior.

teejusb replied:
I thinnnnnnk this is fine. I can't really think of a scenario where it'd be particularly breaking.

teejusb replied:
I think this is probably fine as well. Can't think of a huge downside do it.

joshhead commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
+      curValues.oldest = (curValues.oldest + 1) % max_size;
+    }
+    if (curValues.kCurValues.length === 1) {
+      // WebSocket is considered ready after it gets its first "values" message.
```
I decided to wait for the server to send a sensor reading before loading the rest of the UI, instead of defaulting to 0's. It would be straightforward to switch back the "initialize to 0's" behavior too.

teejusb replied:
I think switching back to to the previous behavior is better only really for one reason.

I'm working on adding a config page with a dropdown that lets you select the serial device to connect to from the UI. Because of this, the UI would need to be loaded before the values message has potentially been sent.

joshhead replied:
That's a great reason, UI selectable serial ports sound very useful.

teejusb commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
-// Amount of panels.
-const num_panels = 4;
+// Maximum number of historical sensor values to retain
+const max_size = 1000;
```
Can we rename to `MAX_SIZE` (even though I know the originally it was `max_size`).

To answer your question in the other comment. The `k` prefix was what's used to denote constants in C++ as part of the Google Style Guide. I copied that paradigm for the JS here, but reading up on Google's style guide for JS, the actually use CONSTANT_CASE, so I'm going to suggest that moving forward:

https://google.github.io/styleguide/jsguide.html#naming-constant-names

I'll run a linter on this in a follow up and fix all the the things I've messed up, but I think tackling the smaller one within this PR sgtm.

joshhead replied:
> Can we rename to MAX_SIZE (even though I know the originally it was max_size).

Will do.

> The k prefix was what's used to denote constants in C++ as part of the Google Style Guide.

Ahh makes sense.

teejusb commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
+// The values are stored in a mutable array in a ref so that they are not
+// subject to the React render cycle, for performance reasons.
+// The Context is to make them easy to access from nested components.
+const CurValuesRefContext = React.createContext();
```
Similarly, I think this can be CUR_VALUES_CONTEXT (or similar).

joshhead replied:
I think MAX_SIZE sounds good, but CUR_VALUES_CONTEXT I'm not so sure about.

From the google style guide you linked, 

> Before choosing constant case, consider whether the field really feels like a deeply immutable constant. For example, if any of that instance's observable state can change, it is almost certainly not a constant. Merely intending to never mutate the object is generally not enough.

This one doesn't feel like a constant to me, since the value of `useContext(CUR_VALUES_CONTEXT)` depends on the value passed to the nearest containing `CUR_VALUES_CONTEXT.Provider`. Also it contains a ref with mutable state.

joshhead replied:
My thought about putting it in a context was, since I was getting rid of some global variables (technically module-level), it would still be easy to access the values/thresholds with a context without passing the props through every layer. Looking at it now though, there aren't that many layers, so I'm thinking I'll remove it anyway. I can put it back if you want to keep it.

teejusb replied:
Yeah I think I agree with you with naming.

Props I think is fine.

teejusb commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
+
+  const curValuesRef = useRef({
+
+    // A history of the past 'max_size' values fetched from the backend.
```
MAX_SIZE

teejusb commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
-      wsCallbacks[action](msg);
+  const emit = useCallback((msg) => {
+    // App should wait for isWsReady to send messages.
+    if (!wsRef.current || !isWsReady) {
```
Do we need `isWsReady` variable? Does the the `wsRef.current.readyState` field suffice?

joshhead replied:
I did it this way for two reasons. One is because I don't actually mark it ready until values start coming in, but I'm changing that back as you requested. The other reason is to make sure `App` re-renders when it changes. Because `isWsReady` is React state, calling `setIsWsReady` will cause the component that called `useState` to re-render. It's less about this particular check in `emit` but I thought it made sense to check the same thing as App.

teejusb commented on webui/src/App.js:
```diff
@@ -479,22 +569,13 @@ function Plot() {
-
+function FSRWebUI(props) {
+  const { curValuesRef, emit, defaults, wsCallbacksRef } = props;
+  const numPanels = defaults.thresholds.length;
```
Can we rename to `numSensors` (that's probably what it should've been named from the start).

teejusb commented on webui/src/App.js:
```diff
@@ -21,86 +21,168 @@ import {
+// Returned `defaults` property will be undefined if the defaults are loading or reloading.
+// Call `reloadDefaults` to clear the defaults and reload from the server.
+function useDefaults() {
+  const [defaults, setDefaults] = useState();
```
Explicitly set this to `undefined`

teejusb commented on webui/src/App.js:
```diff
@@ -299,13 +383,14 @@ function ValueMonitor(props) {
 
-function WebUI() {
+function ValueMonitors(props) {
+  const { emit, numPanels} = props;
```
numSensors

joshhead commented:
@teejusb I addressed the comments and made a couple additional changes. For the exhaustive deps on ValueMonitor, its dependencies are not causing excessive renders at this point (or any re-renders as far as I can see). IMHO ignoring the exhaustive deps warning tends to cause more problems than it solves, but I can rebase to drop that last commit if you want it left alone.

teejusb commented on webui/src/App.js:
```diff
@@ -21,107 +21,187 @@ import {
+  const curThresholds = webUIDataRef.current.curThresholds;
 
-  function EmitValue(val) {
+  const EmitValue = useCallback((val) => {
```
Just for my understanding since I'm not well versed in React, what does `useCallback` here do better instead of defining a function like I had previously?

joshhead replied:
Every render, the ValueMonitor component function runs and EmitValue is created as a new function. Since EmitValue is a dependency of the useEffect call, a new function every render causes the effect to rerun every render. useCallback remembers the function passed to it on the first render and keeps returning that so you have a function with a stable reference (until its own dependencies change).

joshhead replied:
I'm not sure how clear that explanation was but the point is to avoid an infinite render loop.

teejusb replied:
Ahhhh that makes sense! So if `EmitValue` wasn't part of the dependency list, it would be okay to not use `useCallback` as it wouldn't trigger the recursive loop?

joshhead replied:
Right! Actually I was thinking about it and I'm not sure if it actually causes a render loop in this case, but it defeats the purpose of the dependency list. The effect only reruns when its dependencies change, but if you include a dependency that is different every render, it's effectively the same as omitting the dependency list.

teejusb commented:
> @teejusb I addressed the comments and made a couple additional changes. For the exhaustive deps on ValueMonitor, its dependencies are not causing excessive renders at this point (or any re-renders as far as I can see). IMHO ignoring the exhaustive deps warning tends to cause more problems than it solves, but I can rebase to drop that last commit if you want it left alone.

I think my original reasoning for not having the dependencies was because I thought the rendering was handled by `requestAnimationFrame` instead of by React. In my head it worked like:

1) When loading the page, React renders the HTML5 Canvas(es) once
2) The HTML5 canvas re-renders itself using `requestAnimationFrame` with whatever values are found in curValues at some given FPS instead of letting React re-render it.

My understanding was that values are fetched fairly fast (every 10ms iirc), but we don't have to update the plot nearly as fast since we're restricted by the refresh rate of the monitor (e.g. 60Hz for me, or 16.6ms). `requestAnimationFrame` syncs up with the monitor's refresh rate so it made sense to let that handle it.

Honestly I don't know how much this makes sense from an actual webdev standpoint but it made sense in my head. This was the first time I worked with React, and also combining it with HTML5 Canvases so I'm definitely open to learning more about the process.

I think I might need a better understanding as to why adding dependencies is better here as opposed to the previous approach.

teejusb commented:
All of that makes sense! I wasn't quittteeee sure if it was by reference equality or by the contents (but reference does make sense)! All in all the changes look good to me and I'll merge after I get a chance to manually test it myself with my pad. Thanks for this!

teejusb commented:
I tested it and it seemed to work great!

joshhead commented:
Awesome, thanks :)
